src/Cookie.php

```diff
@@ -0,0 +1,494 @@
+    protected $maxAge;
+
+    /**
+     * @var \DateTime|null
+     */
+    protected $expires;
+
+    /**
+     * @var string|null
+     */
+    protected $domain;
+
+    /**
+     * @var string
+     */
+    protected $path;
+
+    /**
+     * @var bool
+     */
+    protected $secure;
+
+    /**
+     * @var bool
+     */
+    protected $httpOnly;
+
+    /**
+     * @param string             $name
+     * @param string|null        $value
+     * @param int|\DateTime|null $expiration
+     * @param string|null        $domain
+     * @param string|null        $path
+     * @param bool               $secure
+     * @param bool               $httpOnly
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        $name,
+        $value = null,
+        $expiration = 0,
+        $domain = null,
+        $path = null,
+        $secure = false,
+        $httpOnly = false
+    ) {
+        $this->validateName($name);
+        $this->validateValue($value);
+
+        $this->name = $name;
+        $this->value = $value;
+        $this->maxAge = is_int($expiration) ? $expiration : null;
+        $this->expires = $this->normalizeExpires($expiration);
+        $this->domain = $this->normalizeDomain($domain);
+        $this->path = $this->normalizePath($path);
+        $this->secure = (bool) $secure;
+        $this->httpOnly = (bool) $httpOnly;
+    }
+
+    /**
+     * Returns the name.
+     *
+     * @return string
+     */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /**
+     * Returns the value.
+     *
+     * @return string|null
+     */
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+    /**
+     * Checks if there is a value.
+     *
+     * @return bool
+     */
+    public function hasValue()
+    {
+        return isset($this->value);
+    }
+
+    /**
+     * Sets the value.
+     *
+     * @param string|null $value
+     *
+     * @return self
+     */
+    public function withValue($value)
+    {
+        $this->validateValue($value);
+
+        $new = clone $this;
+        $new->value = $value;
+
+        return $new;
+    }
+
+    /**
+     * Returns the max age.
+     *
+     * @return int|null
+     */
+    public function getMaxAge()
+    {
+        return $this->maxAge;
+    }
+
+    /**
+     * Checks if there is a max age.
+     *
+     * @return bool
+     */
+    public function hasMaxAge()
+    {
+        return isset($this->maxAge);
+    }
+
+    /**
+     * Sets the max age.
+     *
+     * @param int|null $maxAge
+     *
+     * @return self
+     */
+    public function withMaxAge($maxAge)
+    {
+        $new = clone $this;
+        $new->maxAge = is_int($maxAge) ? $maxAge : null;
+
+        return $new;
+    }
+
+    /**
+     * Sets both the max age and the expires attributes.
+     *
+     * @param int|\DateTime|null $expiration
+     *
+     * @return self
+     */
+    public function withExpiration($expiration)
```

Comment:
ah sorry, i whould have read the the phpdoc. but then i don't get what the idea is. if the param is an int, we use it as max-age, if its a datetime we use it as expires? are we sure that this is a good idea?

i would prefer to have withMaxAge remove the expires property, and vice versa.